tree/dataframe/inc/ROOT/RDFHelpers.hxx

```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
+  fUseShellColours{useShellColours}
+  { }
+
+  ~ProgressHelper() = default;
+
+  // clang-format off
+  /// Thread-safe callback for RDataFrame.
+  /// It will record elapsed times and event statistics, and print a progress bar every second.
+  /// \param slot Ignored.
+  /// \param value Ignored.
+  // clang-format on
+  template<typename T>
+  void operator()(unsigned int /*slot*/, T& value) { operator()(value); }
+
+  // clang-format off
+  /// Thread-safe callback for RDataFrame.
+  /// It will record elapsed times and event statistics, and print a progress bar every second.
+  /// \param value Ignored.
+  // clang-format on
+  template<typename T>
+  void operator()(T& /*value*/) {
+    using namespace std::chrono;
+    // ***************************************************
+    // Warning: Here, everything needs to be thread safe:
+    // ***************************************************
+    fProcessedEvents += fIncrement;
+
+    // We only print every 1s, and only one thread does it. Try to take the lock:
+    if (duration_cast<seconds>(system_clock::now() - fLastPrintTime).count() == 0 || !fPrintMutex.try_lock())
+      return;
+
+    // ***************************************************
+    // Protected by lock from here:
+    // ***************************************************
+    std::lock_guard<std::mutex> lockGuard(fPrintMutex, std::adopt_lock_t{});
+
+    std::size_t eventCount;
+    seconds elapsedSeconds;
+    std::tie(eventCount, elapsedSeconds) = RecordEvtCountAndTime();
+
+    std::cout << "\r";
+    PrintProgressbar(std::cout, eventCount);
+    PrintStats(std::cout, eventCount, elapsedSeconds);
+    std::cout << std::flush;
+  }
+
+
+private:
+  double EvtPerSec() const;
+  std::pair<std::size_t, std::chrono::seconds> RecordEvtCountAndTime();
+  void PrintStats(std::ostream& stream, std::size_t currentEventCount, std::chrono::seconds totalElapsedSeconds) const;
+  void PrintProgressbar(std::ostream& stream, std::size_t currentEventCount) const;
+
+  const std::chrono::time_point<std::chrono::system_clock> fBeginTime = std::chrono::system_clock::now();
+  std::chrono::time_point<std::chrono::system_clock> fLastPrintTime = fBeginTime;
+
+  std::atomic<std::size_t> fProcessedEvents{ 0 };
+  std::size_t fLastProcessedEvents = 0;
+  const std::size_t fMaxEvents;
+  const std::size_t fIncrement;
+
+  std::array<double, 20> fEventsPerSecondStatistics;
+  std::size_t fEventsPerSecondStatisticsIndex = 0;
+
+  const unsigned int fBarWidth;
+
+  std::mutex fPrintMutex;
+  const bool fUseShellColours;
+};
+
+
+// clang-format off
+/// Count events in a tree in the file described by `fileUrl`.
+/// \note This will open and close the file, so don't use this in loops.
+///
+/// \param treename Name of the tree.
+/// \param fileUrl Path or URL to a ROOT file.
+/// \return Event counts. Returns 0 in case of errors.
+// clang-format on
+std::size_t CountEvents(const char* treename, const char* fileUrl);
```

Comment:
I don't see this function being called?

Reply:
Not here, but see code example in PR description. It's in the TODOs that we need to decide how to make this a one-liner.
- [ ] We also need to point out that we are accessing each file *once more*, so maybe extra connections to handle.
- [ ] We also need a solution for chains or trees that come via pointer, e.g. an overload.

Reply:
I see, I suppose that indeed needs some discussion. We should strive for a solution that takes estimates:
- all files have the same amount of entries, open one and scale with the number of files according to RDF
- as the callback is invoked on a newly opened file, update the assumption

Reply:
@eguiraud ^

Reply:
To do something like that we'd need to make the progress bar a core RDF feature (and then add logic to update the guesstimate when a new file is opened, only for datasources that support multiple input files...).

In order to keep things simple, I am fine with just printing `XXX evts/s, YYY evts processed`, without a progress bar, if the total number of events is not provided.

Reply:
Virtually everybody (not using distRDF) will want this feature. Asking users to provide the total event count will trigger the following:
- tutorials will pass the parameter
- analyses will pass the parameter, causing all input files to be opened
- repeat for the next 20 years, with enormous integrated perf losses.

IMHO a progress bar *is* a standard RDF feature, and we need to reduce the damage. I don't think the design you describe is the only option; I could imaging datasources providing abstracted access to "input identifier", "number of entries", such that the progress bar can check whether its datasource identifier has changes, and if so, update the number of entries. Or RDF could construct that progress bar with a callback, or at least with the number of dataset elements - which also sounds like a useful interface for datasources.

Reply:
I was describing requirements more than a design.

What you describe is a different kind of effort than a helper function. Please open a feature request and I'll get to it as soon as possible :)

Reply:
That's missing a point: if we add this, people will use it and it will have detrimental effects on performance. Even though we won't get to doing the better option in a reasonable time we should not add a solution for which we predict detrimental effects, even if that option requires way less effort.

Reply:
I didn't miss that point :)

Reply:
Oh okay sorry - I didn't realize that. I've opened the [feature request](https://github.com/root-project/root/issues/8685) you asked for.